book/src/proposals/booting-from-snapshot.md

```diff
@@ -43,23 +44,70 @@ will return true if `a` is an ancestor of `b`.
+different fork than the one this validiator last rooted. Thus it's critical for
+consistency in `replay_stage` that the order of events when setting a new root
+is:
+
+    1) Write the root to locktower 2) Write the root to blocktree 3) Generate
+    snapshot for that root
+
+3) On startup, the validator checks that the current root in locktower exists
+in blocktree. If not (there was a crash between 2-1 and 2-2), then rewrite the
+root to blocktree.
+
+4) On startup, the validator boots from the snapshot `S`, then replays all
+descendant blocks of `S` that exist in this validator's ledger, building banks
+which are then stored in an output `BankForks`. This is done in
+`blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+5) On startup, the validator calculates the first slot `S_n` from which it is
+not locked out for voting. Every validator persists its locktower state and
+must consult this state in order to boot safely and resume from a snapshot
+without being slashed. From this locktower state and the ancestry information
+embedded in the snapshot, a validiator can derive which banks, are "safe" (See
+the `Determining Vote Safety From a Snapshot and Locktower` section for more
+detals) to vote for.
+
+6) Periodically send canary transactions to the cluster using the validator's
+local recent blockhash.
+
+7) Wait for the following criteria. While waiting, set a new root every time
+2/3 of the cluster's stake roots a bank. This allows the validator to prune its
+state and also calculate leader schedules as it moves across epochs.
+
+   1) Observe a canary transaction in a bank that some threshold of trusted
+   validator stake has voted on. Call this trusted bank `T`.
+   
+   2) The current working bank has slot number `S_current` > `S_n`
+
+8) Start the normal voting process, voting for any slot that satisfies
+`S_current` > `S_n`
+
+
+## Determining `S_n` From a Snapshot and Locktower
 
-Given a `BankForks`, a validator can run some procedure to determine whether it
-can vote on a frozen bank `T` (frozen banks must be present in
-`BankForks.banks`), without violating the lockouts of any vote in `L`. The
-procedure for this is:
+Define the "safety" condition to be: 
 
-1) If `T` < last locktower vote, return false,
+Given a `BankForks`, a validator can run some procedure to determine the first
+slot `S_n` that it can vote on without violating the lockouts of any vote in
+`L`. The procedure for this is:
 
-2) Apply `T` to locktower state `L`. Pop off all votes `S_i` where
-`!is_locked(S_i, T)`. `T` must now be the top of the tower.
+1) Find the earliest vote `V` in the tower for which `is_ancestor(V, S)` is not
+true.
 
-3) If for any remaining vote `L_i` in `L` `!is_ancestor(L_i, T)`, return false.
+2) `V` and every vote after it needs to expire before the validator can vote on
+any descendant of `S`, so `S_n = V + lockout(V)`. If no `V` exists, then `S_n =
```

Comment:
i think in the current implementation of locktower:
If
1) For two votes `L_i `and `L_j` in locktower
2)  `L_i` comes before `L_j` in the tower

There's no guarantee that `L_i + lockout(L_i) >= L_j + lockout(L_j)`

Ths means you actually need to iterate over every `V_i > V` in the tower and take `max(V_i + lockout(V_i)` in order to compute `S_n`